internal/pkg/earpointer/taint.go

```diff
@@ -189,10 +189,73 @@ func IsEARTainted(sink ssa.Instruction, srcRefs ReferenceSet, heap *Partitions,
 	for sink := range sinkedRefs {
 		members := heap.PartitionMembers(sink)
 		for _, m := range members {
-			if _, ok := srcRefs[m]; ok {
+			if srcRefs[m] {
 				return true
 			}
 		}
 	}
 	return false
 }
+
+type SourceSinkTrace struct {
+	Src       *source.Source
+	Sink      ssa.Instruction
+	Callstack []ssa.Call
+}
+
+// Look for the pairs of <source, sink> examining the heap alias information.
+func SourcesToSinks(funcSources source.ResultType, isTaintField func(named *types.Named, index int) bool,
+	heap *Partitions, conf *config.Config) []*SourceSinkTrace {
+
+	var traces []*SourceSinkTrace
+	for fn, sources := range funcSources {
+		// Transitively get the set of functions called within "fn".
+		// This set is used to narrow down the set of references needed to be
+		// considered during EAR heap traversal. It can also help reducing the
+		// false positives and boosting the performance.
+		callees := boundedDepthCallees(fn, conf.EARTaintCallSpan)
+		srcRefs := make(map[*source.Source]ReferenceSet)
+		for _, s := range sources {
+			srcRefs[s] = srcAliasRefs(s, isTaintField, heap, callees)
+		}
+		// Return a source if it can reach the given sink.
+		reachAnySource := func(sink ssa.Instruction) *source.Source {
```

Comment:
If this doesn't _need_ to be dynamically linked, I'd rather it were extracted to a top-level `func` definition with its inputs can be made explicit.